File: src/sample/FieldView.java
package sample;

import javafx.application.Application;
import javafx.event.EventHandler;
import javafx.scene.Scene;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.AnchorPane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Line;
import javafx.scene.shape.Rectangle;
import javafx.stage.Stage;
import sample.field.field_item.*;
import sample.util.FieldUtil;

import java.util.Map;

public class FieldView extends Application {

    private sample.FieldPresenter presenter = new FieldPresenter(this);
    private AnchorPane root;

    @Override
    public void start(Stage primaryStage) throws Exception {

        initialize(primaryStage);
        drawFieldGreed();

    }

    public void drawField(Field field) {
        for (Map.Entry<Position, FieldItem> item : field.getItems().entrySet()) {
            if (item.getValue() instanceof Ship) {
                drawShip(item.getKey());
            } else if (item.getValue() instanceof WoundShip) {
                drawWoundShip(item.getKey());
            } else if (item.getValue() instanceof Miss) {
                drawMiss(item.getKey());
            } else if (item.getValue() instanceof Water) {
                drawWater(item.getKey());
            }
        }
    }

    public void drawMiss(sample.Position position) {
        Rectangle rectangle = new Rectangle(
                FieldUtil.INSTANCE.positionToCoordinationMiddle(position.getCol()) - 4,
                FieldUtil.INSTANCE.positionToCoordinationMiddle(position.getRow()) - 4,
                8,
                8
        );
        root.getChildren().add(rectangle);
    }

    public void drawShip(sample.Position position) {
        Rectangle rectangle = new Rectangle(
                FieldUtil.INSTANCE.positionToCoordination(position.getCol() - 1),
                FieldUtil.INSTANCE.positionToCoordination(position.getRow() - 1),
                FieldUtil.INSTANCE.getITEM_INLINE_COUNT() - 1,
                FieldUtil.INSTANCE.getITEM_INLINE_COUNT() - 1
        );
        rectangle.setFill(Color.GREEN);
        root.getChildren().add(rectangle);
    }

    public void drawWoundShip(sample.Position position) {
        Rectangle rectangle = new Rectangle(
                FieldUtil.INSTANCE.positionToCoordination(position.getCol() - 1),
                FieldUtil.INSTANCE.positionToCoordination(position.getRow() - 1),
                FieldUtil.INSTANCE.getITEM_INLINE_COUNT() - 1,
                FieldUtil.INSTANCE.getITEM_INLINE_COUNT() - 1
        );
        rectangle.setFill(Color.RED);
        root.getChildren().add(rectangle);
    }

    public void drawWater(sample.Position position) {
        Rectangle rectangle = new Rectangle(
                FieldUtil.INSTANCE.positionToCoordination(position.getCol()) + 1,
                FieldUtil.INSTANCE.positionToCoordination(position.getRow()) + 1,
                FieldUtil.INSTANCE.getITEM_INLINE_COUNT() - 1,
                FieldUtil.INSTANCE.getITEM_INLINE_COUNT() - 1
        );
        rectangle.setFill(Color.BLUE);
        root.getChildren().add(rectangle);
    }

    private void drawFieldGreed() {
        for (int lineCounter = 0; lineCounter <= FieldUtil.INSTANCE.getLINE_COUNT(); lineCounter++) {
            double startX = 0;
            double endX = FieldUtil.INSTANCE.getITEM_SIZE() * FieldUtil.INSTANCE.getITEM_INLINE_COUNT();
            double startY = lineCounter * FieldUtil.INSTANCE.getITEM_SIZE();
            double endY = lineCounter * FieldUtil.INSTANCE.getITEM_SIZE();
            Line lineHorizontal = new Line(startX, startY, endX, endY);
            root.getChildren().add(lineHorizontal);
            Line lineVertical = new Line(startY, startX, endY, endX);
            root.getChildren().add(lineVertical);
        }
    }

    private sample.Position getItemNumber(int xPosition, int yPosition) {
        double col = xPosition / FieldUtil.INSTANCE.getITEM_SIZE() + 1;
        double row = yPosition / FieldUtil.INSTANCE.getITEM_SIZE() + 1;
        return new Position((int) col, (int) row);
    }

    private void initialize(Stage primaryStage) {
        root = new AnchorPane();
        primaryStage.setTitle("Hello World");
        primaryStage.setScene(new Scene(root, 1000, 1000));
        primaryStage.show();


        root.addEventFilter(MouseEvent.MOUSE_PRESSED, new EventHandler<MouseEvent>() {
            @Override
            public void handle(MouseEvent event) {
                presenter.shot(getItemNumber((int) event.getX(), (int) event.getY()));
            }
        });
    }

    public static void main(String[] args) {
        launch(args);
    }
}

File: src/sample/field/FieldView.kt
package sample.field

import javafx.scene.layout.AnchorPane
import javafx.scene.paint.Color
import javafx.scene.shape.Line
import javafx.scene.shape.Rectangle
import sample.Position
import sample.field.field_item.*
import sample.util.FieldUtil

class FieldView(var state: FieldState) {

    private val presenter = FieldPresenter(this)
    private lateinit var canvas: AnchorPane
    private var x: Int = 0
    private var y: Int = 0

    fun render(canvas: AnchorPane, x: Int, y: Int) {
        this.x = x
        this.y = y
        this.canvas = canvas
        presenter.start()
        drawFieldGreed()
    }

    fun onClick(x: Int, y: Int) {
        when (state) {
            FieldState.CONSTRUCTOR -> {
                presenter.constructor(FieldUtil.coordinationToPosition(x - this.x, y - this.y))
            }
            FieldState.BATTLE -> {
                presenter.shot(FieldUtil.coordinationToPosition(x - this.x, y - this.y))
            }
        }
    }

    fun drawField(field: Field) {
        for ((key, value) in field.items) {
            when (value) {
                is Ship -> drawShip(key)
                is WoundShip -> drawWoundShip(key)
                is Miss -> drawMiss(key)
                is Water -> drawWater(key)
            }
        }
    }

    fun setShips():Int{
        return presenter.setItems()
    }

    private fun drawMiss(position: sample.Position) {
        val rectangle = Rectangle(
                (FieldUtil.positionToCoordinationMiddle(position.col) - 4).toDouble() + x,
                (FieldUtil.positionToCoordinationMiddle(position.row) - 4).toDouble() + y,
                8.0,
                8.0
        )

        canvas.children.add(rectangle)
    }

    private fun drawShip(position: sample.Position) {
        val rectangle = Rectangle(
                (FieldUtil.positionToCoordination(position.col) + 1).toDouble() + x,
                (FieldUtil.positionToCoordination(position.row) + 1).toDouble() + y,
                (FieldUtil.ITEM_SIZE - 1).toDouble(),
                (FieldUtil.ITEM_SIZE - 1).toDouble()
        )
        rectangle.fill = Color.GREEN
        canvas.children.add(rectangle)
    }

    private fun drawWoundShip(position: sample.Position) {
        val rectangle = Rectangle(
                (FieldUtil.positionToCoordination(position.col) + 1).toDouble() + x,
                (FieldUtil.positionToCoordination(position.row) + 1).toDouble() + y,
                (FieldUtil.ITEM_SIZE - 1).toDouble(),
                (FieldUtil.ITEM_SIZE - 1).toDouble()
        )
        rectangle.fill = Color.RED
        canvas.children.add(rectangle)
    }

    private fun drawWater(position: sample.Position) {
        val rectangle = Rectangle(
                (FieldUtil.positionToCoordination(position.col) + 1).toDouble() + x,
                (FieldUtil.positionToCoordination(position.row) + 1).toDouble() + y,
                (FieldUtil.ITEM_SIZE - 1).toDouble(),
                (FieldUtil.ITEM_SIZE - 1).toDouble()
        )
        rectangle.fill = Color.BLUE
        canvas.children.add(rectangle)
    }

    private fun drawFieldGreed() {
        for (lineCounter in 0..FieldUtil.LINE_COUNT) {
            var startX = 0.0 + x
            var endX = (FieldUtil.ITEM_SIZE * FieldUtil.ITEM_INLINE_COUNT).toDouble() + x
            var startY = (lineCounter * FieldUtil.ITEM_SIZE).toDouble() + y
            var endY = (lineCounter * FieldUtil.ITEM_SIZE).toDouble() + y
            val lineHorizontal = Line(startX, startY, endX, endY)
            canvas.children.add(lineHorizontal)
            startX = (lineCounter * FieldUtil.ITEM_SIZE).toDouble() + x
            endX = (lineCounter * FieldUtil.ITEM_SIZE).toDouble() + x
            startY = 0.0 + y
            endY = (FieldUtil.ITEM_SIZE * FieldUtil.ITEM_INLINE_COUNT).toDouble() + y
            val lineVertical = Line(startX, startY, endX, endY)
            canvas.children.add(lineVertical)
        }
    }


}